### PR TITLE
[imgproc]Add template match with mask.

### PR DESCRIPTION
**Pull request by wangyan42164**
Signed-off-by: Yan Wang <yan.wang@linux.intel.com>

### CODE REVIEW

**Comment by vpisarev** [comment]
@wangyan42164, thanks for the updates and sorry for delay, we've just got back from NY holidays. I mean, the mask in OpenCV is usually treated like this: mask(x,y) != 0 ~ 255, mask(x,y) == 0 ~ 0. It means that mask values should not necessary be all set to 255, they may be 1's or have any other value

**Comment by wangyan42164** [comment]
@vpisarev Thanks for your comments. May you means as the following?
if (mask != 0)
   mask.convertTo(mask, mtype, 1.0 / 255);

if this, I think we may return directly if mask == 0 because the template may be invalid.
Do you agree it?

**Comment by vpisarev** [comment]
@wangyan42164, no, I mean:

<pre>
compare(mask, Scalar::all(0), mask, CMP_NE);
mask.convertTo(mask, mtype, 1.0/255);
</pre>

**Comment by wangyan42164** [comment]
@vpisarev Thanks for your help. I have changed this commit based on your comments. Please review it.

**Comment by vpisarev** [comment]
:+1:

File: modules/imgproc/src/templmatch.cpp [hunk 1]
```diff
@@ -814,12 +814,97 @@ void crossCorr( const Mat& img, const Mat& _templ, Mat& corr,
...
+        Mat corr(corrSize, CV_32F);
+        crossCorr( img2, mask2, corr, corr.size(), corr.type(), Point(0,0), 0, 0 );
+        crossCorr( img, mask_templ, result, result.size(), result.type(), Point(0,0), 0, 0 );
+
+        sqrt(corr, corr);
+        result = result.mul(1/corr);
+        result /= std::sqrt(templSum2);
+    }
+    else
+        CV_Error(Error::StsNotImplemented, "");
```
Comment (Dessix) [inline modules/imgproc/src/templmatch.cpp]: The fact that this does not implement anything but SQDIFF and CCORR_NORMED would be very helpful to have documented.

**Comment by techtonik** [comment]
So is https://github.com/starcraft04/swauto the only documentation for this feature? =)

**Comment by techtonik** [comment]
https://github.com/starcraft04/swauto/blob/095c3703c51ce5136d827f94caf8bb64b65a81ca/functions_opencv.py#L146-L221

**Comment by lhcalibur** [comment]
> Signed-off-by: Yan Wang [yan.wang@linux.intel.com](mailto:yan.wang@linux.intel.com)

Excuse me, i couldn't understand the source code's meaning about matchTemplateMask, may you explain a little, eg which paper or what the equation it use and the formula derivation? Thank you very mutch!

**Pull request merged**